test(broker): add state controller performance tests

npepinpe opened the pull request:
## Description

This PR adds a new performance test to track the `StateController#recover` process, which includes copying a snapshot and opening the underlying RocksDB state. Specifically, it measures the case where the state is considered "large", up to 6GB.

> **Note**
> The way to compute the snapshot size while we're generating it is far from accurate, since RocksDB is working in the background, performing compaction, flushing, etc., so the size is constantly fluctuating. This means that, typically, the final state size is less than the desired targeted size. We could add some factor as a heuristic, but it's unclear to me if that's super useful.
>
> For now, we log the desired and actual snapshot size. I'm a bit worried however that this will significantly skew results, making it hard to assign a correct reference score and maximum deviation.

To simplify adding this test and future tests, a new `JMHTestCase` utility is added, which lets you easily write a JUnit test that runs a benchmark and asserts its result. Additionally, a new `JMHTest` annotation is added, which is coupled with a JUnit 5 extension that will inject a pre-configurd `JMHTestCase` into a test method as a parameter. The annotation ensures all JMH tests will have an appropriate tag and will be excluded from the main CI pipeline, but included in the performance test job.

> **Note**
> All of this can be expanded later as we need more stuff.

There are some notable downsides to this approach of running JMH tests however:

1. Since JMH forks, it's not possible to debug the tests easily locally. To do so, you have to disable forking. Not too difficult, but not obvious when you're not familiar with it.
2. The reference score is typically different locally and on CI, and we already see that it sometimes randomly fluctuate on CI. Would be nice to find a better approach.
3. When a benchmark fails with an error, the actual assertion error is no benchmarks were run. This is pretty bad; luckily, the benchmark logs are present in the test output, but it means you have to wait for the complete workflow to finish before being able to grab them.

## Related issues

related #13775 

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/JMHStatePerformanceTest.java:
```diff
@@ -0,0 +1,63 @@
+    }
+  }
+
+  private static void assertDeviationWithin(
+      final RunResult result, final double referenceScore, final double maxDeviation) {
+    final double score = result.getPrimaryResult().getScore();
+    final double deviation = Math.abs(score / referenceScore - 1);
+
+    Assertions.assertThat(deviation)
+        .withFailMessage(
+            () -> {
+              final String deviationString = DECIMAL_FORMAT.format(deviation * 100) + "%";
+              final String maxDeviationString = DECIMAL_FORMAT.format(maxDeviation * 100) + "%";
+
+              return String.format(
+                  "Expected reference score is '%.2f' got '%.2f', deviation %s exceeds maximum allowed deviation %s",
+                  REFERENCE_SCORE, score, deviationString, maxDeviationString);
+            })
+        .isLessThan(maxDeviation);
```
🔧 Since it is the same code as on the previous JMH test I'm wondering whether we should make a helper out of it.

npepinpe replied:
Alright, I can extract it to a helper. So I see two things:

1. A JMH test case utility - pass in your target JMH class, and it spits out a runnable or something easy to use.
2. A JMH assert to easily assert target deviation (i.e. this method).

Maybe combine them both, since it's fairly small.

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/JMHStatePerformanceTest.java:
```diff
@@ -0,0 +1,63 @@
+
+    // then
+    for (final RunResult runResult : runResults) {
+      // we are fine with around 10 percent deviation but not more
```
```suggestion
      // we are fine with some deviation but not more than the max
```

Needs to be fixed on the other class as well :)

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/TestState.java:
```diff
@@ -0,0 +1,130 @@
+  @SuppressWarnings("resource")
+  private ConstructableSnapshotStore createSnapshotStore(
+      final FileBasedSnapshotStoreFactory storeFactory, final Path tempDirectory) {
+    storeFactory.createReceivableSnapshotStore(tempDirectory, 1);
```
❓ Why we need this?

npepinpe replied:
There's no method to create a constructable snapshot store :smile: You can only create a receivable one (which will submit the actor as well), then grab the constructable one from the factory.

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/TestState.java:
```diff
@@ -0,0 +1,130 @@
+
+    final var storeFactory = new FileBasedSnapshotStoreFactory(actorScheduler, 1);
+    final var snapshotStore = createSnapshotStore(storeFactory, tempDirectory);
+    takeSnapshot(snapshotStore, keyCount, keySize, valueSize);
```
🔧 TakeSnapshot confused me a little. What we actually do is to write our ZeebeDb to that path, but we don't take a snapshot or checkpoint in that sense from the generated data. Might make sense to either split this up or rename it.

npepinpe replied:
Since the controller recovers from the snapshot store, this method actually does take a snapshot (via snapshot.persist()), but I get what you mean. It's more like generating the snapshot.

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/TestState.java:
```diff
@@ -0,0 +1,130 @@
+      final int keySize,
+      final int valueSize) {
+    final var snapshot = snapshotStore.newTransientSnapshot(1, 1, 1, 1).get();
+    snapshot.take(path -> takeSnapshot(path, keyCount, keySize, valueSize)).join();
+    snapshot.persist().join();
```
❓ I guess we need to use these methods in order to correctly recover, since it creates checksums etc?

npepinpe replied:
Correct, and it makes sure the snapshot store sees it as the latest snapshot, and its in memory state is set up properly.

Zelldon commented on .github/workflows/ci.yml:
```diff
@@ -267,6 +267,9 @@ jobs:
           | tee "${BUILD_OUTPUT_FILE_PATH}"
+        env:
+          # creates a state of roughly 4GB
+          STATE_CONTROLLER_PERFORMANCE_TESTS_KEYCOUNT: "10000"
```
💭 I was thinking what we actually care about is the end size of the state, like "I want to test the performance for a 10 gig snapshot." Would be cool if we could configure this in that way, instead of the key count since this is a bit abstract and not really clear what the end result is. or what do you think?

npepinpe replied:
Yeah, being able to have a target size would be best. I struggled a little with this because RocksDB flushes asynchronously, and the only real way to have a "final" DB size is by closing it :smile: (maybe? Maybe I missed something). But yeah, it'd be nicer to target a size and nothing else :thinking:

Zelldon reviewed (approved):
Great! Thanks @npepinpe it is cool that we add more of such tests 👍🏼 I had just some smaller comments or thoughts.

npepinpe commented:
With https://github.com/camunda/zeebe/pull/13884/commits/18bcf57c09d6a2569adb3d7c72d12eae8e54c97a, I added some utilities to easily create junit tests which run JMH benchmarks.

You can see examples in https://github.com/camunda/zeebe/pull/13884/commits/c4c061f674f6d48e50bf27c7ef235c95bf8b4e2c

While I moved the test method to the same class as the benchmark, it can still be moved away, and there's some arguments to be made for keeping them separate for sure. I'm not strongly for or against, let me know what you think. We can still get rid of the JUnit extension and use only the utilities, but imo the extension is nice to ensure all of the tests share the same tag, and as such as included in the appropriate job (and still runnable locally).

Zelldon commented:
Will take a look tomorrow

npepinpe commented:
Another downside right now, the performance test job is quite slow. Since broker depends on the engine, the tests will always run sequentially. The engine test takes 6 minutes, and now the broker one takes 3 minutes. That's 9 minutes alone, so it's often the last job to finish :sweat_smile:

npepinpe commented:
One way to get a more accurate snapshot size is to pause and resume background work in RocksDB manually. However that kind of breaks our ZeebeDb abstraction (maybe?)

npepinpe commented:
Yeah, the way I set up building the state to a certain size is too inaccurate, leading to the variation being too high. :thinking:

npepinpe commented:
I see a few options:

1. Allow pausing/resuming background work for the DB. This will ensure we compute proper sizes without having to worry about compaction/flushing. Not too keen on this because it's modifying the production API to suit tests =/
2. Close the DB between batch inserts, which will give us an idea of the "final" size. This is a bit better, but increases the startup time. This is my preferred approach - it's clean and simple, and it's really only inefficient once, when the state is being set up.
3. Instead of asserting the deviation, we assert that the score is above a minimum target score. This risks missing future improvements, but since what we're testing is fairly small in scope, it might be fine.

npepinpe commented:
Hm, in that last run we only created a 3GB state instead of 4. Unclear how this happened, since we measure the size of the snapshot _after_ closing the DB...

:weary:

npepinpe commented:
So it keeps fluctuating between 0.5 and 0.65. That's pretty big deviation, and this time the state size is pretty stable, so I don't know where it is. That said, it's a relatively big deviation, but in absolute terms, it means a difference of 300ms, which isn't crazy high.

Zelldon commented on test-util/src/main/java/io/camunda/zeebe/test/util/jmh/JMHAssert.java:
```diff
@@ -0,0 +1,61 @@
+import org.openjdk.jmh.results.RunResult;
+
+/** Convenience class to assert results of a JMH test (via {@link RunResult}. */
+public final class JMHAssert extends AbstractAssert<JMHAssert, RunResult> {
```
👍🏼 Mmmh nice..

Zelldon commented on test-util/src/main/java/io/camunda/zeebe/test/util/jmh/JMHTestCase.java:
```diff
@@ -0,0 +1,114 @@
+public final class JMHTestCase {
+  private final ChainedOptionsBuilder builder;
+
+  public JMHTestCase(final ChainedOptionsBuilder builder) {
```
🔧 I guess we can make the ctor private then

npepinpe replied:
I kept it up in case someone wanted to pass some weirdly custom thing, but I guess it can be customized via `withOptions` anyway, so :+1:

Zelldon commented on test-util/src/main/java/io/camunda/zeebe/test/util/jmh/JMHTestCase.java:
```diff
@@ -0,0 +1,114 @@
+ * &#64;Test
+ * void shouldBenchmark() {
+ *   // given
+ *   final var testCase = JMHTestCase.of(MyBenchmarkClass.class, "measurePropertyFoo");
+ *
+ *   // when
+ *   final var assertResult = testCase.run();
+ *
+ *   // then
+ *   assertResult.isWithinDeviation(3.01, 0.2);
```
🚀

Zelldon commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/LargeStateControllerPerformanceTest.java:
```diff
@@ -61,18 +66,30 @@ public void setup() throws Throwable {
+  void shouldRecoverStateWithinDeviation(final JMHTestCase testCase) {
+    // given - when - then
+    testCase.run().isWithinDeviation(3.5, 0.15);
+  }
```
🔧 I would prefer to pull the statements into separate sections, so it is a bit more clearer.

Like:

```
// given

// when
final var result = testCase.run();

// then
restult.isWithinDeviation(3.5, 0.15);
```

Zelldon commented on engine/src/test/java/io/camunda/zeebe/engine/perf/EngineLargeStatePerformanceTest.java:
```diff
@@ -128,4 +130,10 @@ public Record<?> measureProcessExecutionTime() {
+  void shouldProcessWithinExpectedDeviation(final JMHTestCase testCase) {
+    // given - when - then
+    testCase.run().isWithinDeviation(450, 0.15);
+  }
```
🔧 Same here as above :)

Zelldon commented on .github/workflows/ci.yml:
```diff
@@ -268,8 +268,7 @@ jobs:
         env:
-          # creates a state of roughly 4GB
-          STATE_CONTROLLER_PERFORMANCE_TESTS_KEYCOUNT: "10000"
+          LARGE_STATE_CONTROLLER_PERFORMANCE_TEST_SIZE_GB: "4"
```
👍🏼 Thanks for doing this 🤗

Zelldon reviewed (approved):
Great! Thanks @npepinpe, I had smaller comments but I think we are in a pretty good state.

Is the test still flaky?

npepinpe commented on broker/src/test/java/io/camunda/zeebe/broker/system/partitions/impl/perf/TestState.java:
```diff
@@ -80,39 +80,55 @@ private ZeebeRocksDbFactory<ZbColumnFamilies> createDbFactory() {
   }
 
+  private static long computeSnapshotSize(final Path root) {
+    try (final var files = Files.walk(root)) {
```
The property is (according to their wiki) an estimate of the live SST data. I have no idea how accurate it is, that's why I went with checking the file size :shrug: Since it includes L0, I expect it would also fluctuate just as much if we don't stop background work somehow.

npepinpe replied:
Good point about taking a checkpoint, but here we're not taking a checkpoint, so it shouldn't have any impact.

npepinpe commented:
With a 20% max deviation it's not flaky, as it falls within these bounds. Though considering the engine is sometimes flaky, I'm also not 100% confident.

Like I suggested above, it might be that we want to take a different approach than measuring some expected deviation from a reference score. For example, we could only assert that we meet a specific target (e.g. score is greater than some fixed lower bound). Downside here is that thresholds will be different everywhere it's being run, so it may pass locally, but fail in CI. I guess that's still the case with the max deviation though :sweat_smile:

npepinpe commented:
The only remaining thing imo is that the job is the slowest of all now :smile: And the more tests we add, the slower it's going to get. We should eventually look into speeding it up, since each test takes minutes to execute, and likely get executed sequentially...

npepinpe commented:
bors merge
